nexus/db-queries/src/db/datastore/address_lot.rs

```diff
@@ -0,0 +1,333 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use super::DataStore;
+use crate::context::OpContext;
+use crate::db;
+use crate::db::datastore::PgConnection;
+use crate::db::error::public_error_from_diesel_pool;
+use crate::db::error::ErrorHandler;
+use crate::db::error::TransactionError;
+use crate::db::model::Name;
+use crate::db::model::{AddressLot, AddressLotBlock, AddressLotRsvdBlock};
+use crate::db::pagination::paginated;
+use async_bb8_diesel::{
+    AsyncConnection, AsyncRunQueryDsl, Connection, ConnectionError, PoolError,
+};
+use chrono::Utc;
+use diesel::result::Error as DieselError;
+use diesel::{ExpressionMethods, QueryDsl, SelectableHelper};
+use diesel_dtrace::DTraceConnection;
+use ipnetwork::IpNetwork;
+use nexus_types::external_api::params;
+use nexus_types::identity::Resource;
+use omicron_common::api::external::http_pagination::PaginatedBy;
+use omicron_common::api::external::{
+    CreateResult, DataPageParams, DeleteResult, Error, ListResultVec,
+    LookupType, NameOrId, ResourceType,
+};
+use ref_cast::RefCast;
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct AddressLotCreateResult {
+    pub lot: AddressLot,
+    pub blocks: Vec<AddressLotBlock>,
+}
+
+impl DataStore {
+    pub async fn address_lot_create(
+        &self,
+        opctx: &OpContext,
+        params: &params::AddressLotCreate,
+    ) -> CreateResult<AddressLotCreateResult> {
+        use db::schema::address_lot::dsl as lot_dsl;
+        use db::schema::address_lot_block::dsl as block_dsl;
+
+        self.pool_authorized(opctx)
+            .await?
+            .transaction_async(|conn| async move {
```

Comment:
I still agree explicit, round-trip transactions are the way to go in the interest of time. I think it would be useful to file an issue that we should audit these, and potentially include metrics around them. I don't really have a sense of how likely these are to be a pain point, but an issue would at least give us a place to collect data or other problems we might hit.

Reply:
Done
- https://github.com/oxidecomputer/omicron/issues/2811